design-proposals/primary_network_binding_for_seamless_migration.md

```diff
@@ -0,0 +1,596 @@
+
+> **Note**: Consuming a CNI IPAM requirement for the primary pod network
+> is not a hard requirement. There are exceptions to this, e.g. instead
+> of consuming the CNI IPAM, the guest can consume an IP from the
+> network-provider, set statically through cloud-init or other means.
+
+### Migration Support
+
+Kubevirt allows a VM to migrate only if it uses the masquerade binding
+or a plugin binding that is marked
+as [supporting migration](https://kubevirt.io/user-guide/network/network_binding_plugins/#migration)
+for its primary pod network (or if the primary is not used at all).
+The ability to migrate is agnostic to the support for seamless migration
+with no traffic degradation. In fact, the primary pod network traffic is
+currently expected to get broken during the migration (e.g. with
+masquerade, the pod IP change/officials).
+
+Users are recommended to use the secondary network with a L2
+connectivity to preserve network connectivity with minimal impact. This
+also comes with a limitation of not attempting to manage IP on secondary
+networks, due to dependencies on sticky IP allocations for the pods up
+to the ability to pass the IP allocation down to the VM itself.
+
+The proposed solutions need to take into account the challenges the
+existing known bindings have with the primary pod network and with migration.
+
+## Proposal
+
+The following proposal attempts to provide a solution from VMs that use
+the primary pod network for the majority of the traffic and require a
+seamless migration with no (or minimal) traffic connectivity
+interruption.
+
+The solution should consider two important points:
+
+- Be agnostic to a specific network provided and CNI, by defining the
+  requirements well.
+- Co-exit with the current solutions without adding too much burden on
+  maintenance. E.g. the logic and codebase should not diverge too much
+  from the existing core network flow.
+
+### Pre Requirements
+
+The main challenge when working with the primary pod network is to
+assure connectivity recovers after a VM migration. A new requirement is
+added with this proposal, to preserve existing connectivity (e.g. TCP
+connection) with minimal (to none) disruption.
+
+As we are focusing here on the primary pod network, these requirements
+translate to several pre-requirements from the network provider & CNI:
+
+- Sticky IP: An IPAM plugin assigns an IP for each pod. In order to
+  support migration and preserve seamless connectivity, the pod IP on
+  the target node must use the same IP.
+  The network-provider/CNI/IPAM needs to assure such an option exists so
+  the same IP used for the primary pod network can be moved to the
+  target pod.
+- Flat Cluster Subnet: In order to allow the same IP to be assigned on a
+  different node when migration occurs, only a single subnet needs to
+  exist from the entire cluster on the primary pod network.
+- Traffic hand-off on migration: With the need to keep the same IP on
```

Comment:
>   same IP on both the source and target nodes/pods, the network-provider needs to
>   assure traffic is correctly directed to the active VM. I.e.during the
>   migration process, the source VM is the active one (which may take
>   several tens of seconds), afterward when the hand-off is moved to the target

This expresses that the two pods co-exist. If you prefer some other clearer wording, could you please suggest something?

> and that the switch is also a problem.

This is the network-provider business, not ours. I mean, the pre-requirment is for it to assure traffic flows correctly to the two pods. I tried to avoid using specifics as much as possible, just because I do not want to hint about any specific implementation.
I do know that for SDN providers (e.g. OVN), they can duplicate traffic to both ports.